File: plugins/cuebook/skills/render-cuebook-thesis-chart/scripts/audit_chart_svg.py
#!/usr/bin/env python3
"""Audit a rendered Cuebook chart for public-style leakage and design locks."""

from __future__ import annotations

import argparse
import json
import sys
import xml.etree.ElementTree as ET
from pathlib import Path
from typing import Any


def finding(code: str, message: str) -> dict[str, str]:
    return {"code": code, "message": message}


def local_name(tag: str) -> str:
    return tag.rsplit("}", 1)[-1]


def audit(path: Path) -> dict[str, Any]:
    errors: list[dict[str, str]] = []
    warnings: list[dict[str, str]] = []
    try:
        root = ET.parse(path).getroot()
    except (OSError, ET.ParseError) as exc:
        return {"valid": False, "errors": [finding("SVG_READ", str(exc))], "warnings": []}
    if local_name(root.tag) != "svg":
        errors.append(finding("SVG_ROOT", "Root element must be svg."))

    profile = root.attrib.get("data-style-profile")
    if profile not in {"cuebook_feed_v1", "cuebook_detail_v1"}:
        errors.append(finding("STYLE_PROFILE", "Rendered SVG must declare a Cuebook style profile."))
    if root.attrib.get("font-variant-numeric") != "tabular-nums":
        errors.append(finding("TABULAR_NUMS", "Market charts must use tabular numerals."))
    if root.attrib.get("letter-spacing") != "0":
        errors.append(finding("LETTER_SPACING", "Cuebook chart letter spacing must remain zero."))

    elements = list(root.iter())
    names = {local_name(item.tag) for item in elements}
    for banned in {"linearGradient", "radialGradient", "filter"} & names:
        errors.append(finding("DECORATIVE_EFFECT", f"Cuebook charts do not use {banned}."))

    all_text = " ".join("".join(item.itertext()) for item in elements if local_name(item.tag) in {"title", "desc", "text"})
    if "Cuebook" not in all_text:
        errors.append(finding("WATERMARK", "Cuebook watermark text is missing."))

    public_titles = [item for item in elements if local_name(item.tag) == "text" and item.attrib.get("id") == "public-title"]
    if len(public_titles) != 1:
        errors.append(finding("PUBLIC_TITLE", "Rendered chart must contain one visible public title."))
    else:
        lines = [item for item in public_titles[0] if local_name(item.tag) == "tspan"]
        if len(lines) > 2:
            errors.append(finding("TITLE_LINES", "Feed/detail chart title may use at most two lines."))

    if profile == "cuebook_feed_v1":
        banned_phrases = {
            "CONDITIONAL": "internal artifact state",
            "DRAFT": "internal artifact state",
            "Cuebook OHLCV": "provenance footer",
            "结算条件": "settlement prose panel",
            "Settlement ·": "settlement prose panel",
            "Solid candle": "rendering guide",
            "hollow/dashed": "rendering guide",
            "实体/实线": "rendering guide",
            "从观点描述中提取": "workflow narration",
            "Cuebook 从观点": "workflow narration",
            "schema_version": "schema metadata",
        }
        for phrase, meaning in banned_phrases.items():
            if phrase.lower() in all_text.lower():
                errors.append(finding("FEED_LEAKAGE", f"Feed SVG exposes {meaning}: {phrase}."))
        width = float(root.attrib.get("width", 0))
        height = float(root.attrib.get("height", 0))
        if width / max(height, 1) < 1.45:
            warnings.append(finding("FEED_ASPECT", "Feed chart is unusually tall; inspect it at thumbnail size."))

    return {
        "valid": not errors,
        "style_profile": profile,
        "errors": errors,
        "warnings": warnings,
    }


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("path", type=Path)
    args = parser.parse_args()
    result = audit(args.path)
    print(json.dumps(result, ensure_ascii=False, indent=2))
    return 0 if result["valid"] else 1


if __name__ == "__main__":
    sys.exit(main())

File: plugins/cuebook/skills/render-cuebook-thesis-chart/scripts/audit_chart_svg.mjs
#!/usr/bin/env node
// Audit a rendered Cuebook chart for public-style leakage and design locks.

import { readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

export const finding = (code, message) => ({ code, message });
export const local_name = (tag) => String(tag).split("}").at(-1);

function attrs(source) {
  const result = {};
  for (const match of source.matchAll(/([:\w-]+)\s*=\s*(["'])(.*?)\2/gs)) result[match[1]] = match[3];
  return result;
}

function textContent(source) {
  return source.replace(/<[^>]*>/g, " ")
    .replaceAll("&lt;", "<").replaceAll("&gt;", ">").replaceAll("&amp;", "&")
    .replaceAll("&quot;", '"').replaceAll("&apos;", "'");
}

export function audit(file) {
  const errors = [];
  const warnings = [];
  let source;
  try { source = readFileSync(file, "utf8"); }
  catch (error) { return { valid: false, errors: [finding("SVG_READ", error.message)], warnings: [] }; }
  const rootMatch = /^\s*(?:<\?xml[\s\S]*?\?>\s*)?(?:<!--[^]*?-->\s*)?<([\w:.-]+)\b([^>]*)>/i.exec(source);
  if (!rootMatch || !source.includes("</")) return { valid: false, errors: [finding("SVG_READ", "no element found")], warnings: [] };
  const rootName = rootMatch[1].split(":").at(-1);
  const rootAttrs = attrs(rootMatch[2]);
  if (rootName !== "svg") errors.push(finding("SVG_ROOT", "Root element must be svg."));

  const profile = rootAttrs["data-style-profile"];
  if (!new Set(["cuebook_feed_v1", "cuebook_detail_v1"]).has(profile)) errors.push(finding("STYLE_PROFILE", "Rendered SVG must declare a Cuebook style profile."));
  if (rootAttrs["font-variant-numeric"] !== "tabular-nums") errors.push(finding("TABULAR_NUMS", "Market charts must use tabular numerals."));
  if (rootAttrs["letter-spacing"] !== "0") errors.push(finding("LETTER_SPACING", "Cuebook chart letter spacing must remain zero."));

  const names = new Set([...source.matchAll(/<\/?([\w:.-]+)/g)].map((match) => match[1].split(":").at(-1)));
  for (const banned of ["linearGradient", "radialGradient", "filter"]) {
    if (names.has(banned)) errors.push(finding("DECORATIVE_EFFECT", `Cuebook charts do not use ${banned}.`));
  }
  const allText = [...source.matchAll(/<(title|desc|text)\b[^>]*>([\s\S]*?)<\/\1>/gi)].map((match) => textContent(match[2])).join(" ");
  if (!allText.includes("Cuebook")) errors.push(finding("WATERMARK", "Cuebook watermark text is missing."));

  const publicTitles = [...source.matchAll(/<text\b([^>]*)>([\s\S]*?)<\/text>/gi)].filter((match) => attrs(match[1]).id === "public-title");
  if (publicTitles.length !== 1) errors.push(finding("PUBLIC_TITLE", "Rendered chart must contain one visible public title."));
  else if ([...publicTitles[0][2].matchAll(/<tspan\b/gi)].length > 2) errors.push(finding("TITLE_LINES", "Feed/detail chart title may use at most two lines."));

  if (profile === "cuebook_feed_v1") {
    const bannedPhrases = new Map([
      ["CONDITIONAL", "internal artifact state"], ["DRAFT", "internal artifact state"], ["Cuebook OHLCV", "provenance footer"],
      ["\u7ed3\u7b97\u6761\u4ef6", "settlement prose panel"], ["Settlement ·", "settlement prose panel"], ["Solid candle", "rendering guide"],
      ["hollow/dashed", "rendering guide"], ["\u5b9e\u4f53/\u5b9e\u7ebf", "rendering guide"], ["\u4ece\u89c2\u70b9\u63cf\u8ff0\u4e2d\u63d0\u53d6", "workflow narration"],
      ["Cuebook \u4ece\u89c2\u70b9", "workflow narration"], ["schema_version", "schema metadata"],
    ]);
    for (const [phrase, meaning] of bannedPhrases) {
      if (allText.toLowerCase().includes(phrase.toLowerCase())) errors.push(finding("FEED_LEAKAGE", `Feed SVG exposes ${meaning}: ${phrase}.`));
    }
    const width = Number(rootAttrs.width ?? 0);
    const height = Number(rootAttrs.height ?? 0);
    if (width / Math.max(height, 1) < 1.45) warnings.push(finding("FEED_ASPECT", "Feed chart is unusually tall; inspect it at thumbnail size."));
  }
  return { valid: errors.length === 0, style_profile: profile ?? null, errors, warnings };
}

function main(argv) {
  if (argv.length !== 1) {
    process.stderr.write("usage: audit_chart_svg.mjs path\n");
    return 2;
  }
  const result = audit(argv[0]);
  process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);
  return result.valid ? 0 : 1;
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) process.exit(main(process.argv.slice(2)));
